foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/asm/ASMFactory.java

```diff
@@ -41,8 +41,8 @@ public class ASMFactory {
     private final static String ASM_ECLIPSELINK_CLASS_VISITOR = "org.eclipse.persistence.internal.libraries.asm.ClassVisitor";
 
     //Should be changed in case of ASM upgrade
-    public final static int ASM_API_SELECTED = Opcodes.valueInt("ASM9");
```

Comment:
@anija-anil The test failures are caused by a static initialization ordering problem if the Opcodes class is initialized before ASMFactory.  When ASMFactory is initialized first there is no issue; when Opcodes is initialized first, it has a reference to ASMFactory which triggers ASMFactory to then initialize before Opcodes completes initialization and ASM_API_SELECTED ends up set to 0. To avoid the circular initialization error, I'd recommend moving lines 38 and 39 above to Opcodes, and then replacing them here with:

    public final static String ASM_SERVICE_ECLIPSELINK = Opcodes.ASM_SERVICE_ECLIPSELINK;
    public final static String ASM_SERVICE_OW2 = Opcodes.ASM_SERVICE_OW2;

Reply:
Resolved